module/zfs/zio.c

```diff
@@ -3245,16 +3245,37 @@ zio_ddt_write(zio_t *zio)
 	ddt_t *ddt = ddt_select(spa, bp);
 	ddt_entry_t *dde;
 	ddt_phys_t *ddp;
+	boolean_t added = B_FALSE;
 
 	ASSERT(BP_GET_DEDUP(bp));
 	ASSERT(BP_GET_CHECKSUM(bp) == zp->zp_checksum);
 	ASSERT(BP_IS_HOLE(bp) || zio->io_bp_override);
 	ASSERT(!(zio->io_bp_override && (zio->io_flags & ZIO_FLAG_RAW)));
 
 	ddt_enter(ddt);
-	dde = ddt_lookup(ddt, bp, B_TRUE);
+	if (spa->spa_dedup_max_size &&
+	    ddt_get_ddt_dsize(spa) >= spa->spa_dedup_max_size) {
+		/* Over the desired DDT size, set 'nogrow' flag */
+		dde = ddt_lookup(ddt, bp, B_TRUE, NULL);
+	} else {
+		dde = ddt_lookup(ddt, bp, B_FALSE, &added);
+	}
+	if (dde == NULL) {
+		zp->zp_dedup = B_FALSE;
+		BP_SET_DEDUP(bp, B_FALSE);
+		ASSERT(!BP_GET_DEDUP(bp));
+		zio->io_pipeline = ZIO_WRITE_PIPELINE;
+		ddt_exit(ddt);
+		return (zio);
+	}
+
 	ddp = &dde->dde_phys[p];
 
+	if (added) {
+		if (spa->spa_ddt_dsize != ~0ULL)
+			spa->spa_ddt_dsize += ddt_entry_size();
```

Comment:
It's hard to predict the impact on the on-disk size, given the ZAP and compression. (Also ditto blocks are easy to predict but not taken into account here.)  The impact is unlikely to be exactly `sizeof (struct ddt_phys)`.

What problem is this trying to solve?

Reply:
This just grows the cached size by the sizeof() to prevent going over the limit before the next ddt_sync() where we will calculate the actual size. It is just short term to avoid blowing too far past the quota

Reply:
Rather than tying this to `sizeof (struct ddt_phys)`, I would suggest adding a #define for the "fake space" we add for each "pending" addition to the DDT.

Reply:
That would seem to complicate the code significantly -- it would have to keep an entirely different accounting "pending additions," since that would then have to be decremented as well if reference counts dropped, unless I'm not understanding what you're saying here.

Reply:
> that would then have to be decremented as well if reference counts dropped

If a refcount goes to zero, then we will remove an entry from the DDT.  I think you're saying that you want to keep track of that so that these "pending removals" can balance out "pending additions".  That's an interesting extension on top of what you're doing here, but I wouldn't say that you "have to" do that.

My suggestion is that when we know we'll be adding an entry, instead of doing `spa->spa_ddt_dsize += sizeof (struct ddt_phys)`, you do `spa->spa_ddt_pending_additions++`, and then when checking the limit, do something like `if (ddt_dsize + pending_additions * DDE_ESTIMATED_SIZE > ddt_quota)`.  When we finish syncing a txg, you'd set `spa_ddt_pending_additions=0`, which is equivalent to what you're doing now by setting `spa_ddt_dsize` back to the on-disk size at the end of each txg.  I am guessing this will be a net change of less than a dozen additional lines of code from what you're proposing here.

Reply:
I agree that this will make it more obvious to the reader what is going on

Reply:
Hm, I don't think it'll make it clearer, honestly, but it's Alan's choice here :).

I am still not sure what "DDE_ESTIMATED_SIZE" would be.  I *really* don't like picking some random number; I like sizeof because if anything changes, so will it.  So if you've got a suggestion for some math for it, I'm happy to change that. :)

Reply:
Again, I don't think this makes it more obvious, at all, what is going on.  Because this code:
`        if (spa->spa_dedup_table_quota &&
            ddt_get_ddt_dsize(spa) >= spa->spa_dedup_table_quota)
`

changes to `
        if (spa->spa_dedup_table_quota &&
            (ddt_get_ddt_dsize(spa) + spa->ddt_pending * ddt_entry_size())
            >= spa->spa_dedup_table_quota))
`

(that's in ddt_check_overquota()).  I really have to ask:  how is that clearer or more readable?

The code in zio_ddt_write() changes to
`
                        spa->spa_ddt_pending++;
`
instead of
`
			spa->spa_ddt_dsize += ddt_entry_size();
`

The only way I can see that it's clearer (at the cost of some other obfuscation) is that it's not modifying the DDT disk size.  That may be what you meant, in which case, ok, I can see that, as it goes from being an estimate of the new disk size to the actual disk size.

Reply:
> how is that clearer or more readable?

The improvement is in the conceptual model, not necessarily to the number of lines of code.  I'm suggesting that we separate these three ideas, which are currently all combined into `spa_ddt_dsize`):
1. "how much space is used by the dedup table on disk" (`spa_ddt_dsize`), which is an actual fact.
2. "how many entries will we be adding to the dedup table this txg", (`spa_ddt_pending_additions`), which is an actual fact
3. "what is the estimate of how much disk space will be used after the currently-pending modifications", which is necessarily going to be a very rough guess

>I am still not sure what "DDE_ESTIMATED_SIZE" would be. I really don't like picking some random number; I like sizeof because if anything changes, so will it.

I agree, I'm not sure what the value of this macro should be.  The problem with the current proposal is that it implies that `sizeof(struct ddt_phys)` is the amount of space that will be used on disk by an entry in the DDT, but this is not the case.  While that struct is a component of the on-disk representation, there are 4 versions of it, it is compressed twice, triple-dittoed, chunk-ified by the ZAP.  Plus the on-disk dedup entry includes the `ddt_key_t`.  `ddt_zap_update()` is where the ddt_entry_t begins to be transformed into the on-disk representation.  After all that, the average on-disk size may be 64, bytes, but if so its relationship to `sizeof(struct ddt_phys)` is purely coincidental. (But I'm pretty sure it's much more than that, since we have 32 bytes of incompressible checksum, times 3 ditto blocks = 96 bytes just for the checksum.)

I would suggest that you determine the average ZAP entry size experimentally, and hard-code it in a macro, with a comment explaining that it's a rough guess.